### PR TITLE
[ML-DataFrame] stop task at completion of data frame function

### PR DESCRIPTION
**Pull request by hendrikmuhs**
auto stops and removes the persistent task after the indexer has finished

### CODE REVIEW

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java [hunk 1]
```diff
@@ -584,6 +584,14 @@ protected void doSaveState(IndexerState indexerState, Map<String, Object> positi
                                 next.run();
                             }
                     ));
+
+            // TODO: check whether continuous data frames is enabled when available
+            if (indexerState.equals(IndexerState.STARTED) && transformTask.currentCheckpoint.get() == 1) {
+                auditor.info(transformConfig.getId(), "Stopping data frame transform and removing task");
+                logger.info("Stopping data frame transform [{}] and removing task", transformConfig.getId());
+                transformTask.shutdown();
```
Comment (benwtrent) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: The last state saved in the index will be `STARTED` and there are no guarantees that the last state will be persisted by the time we stop this task. It seems to me that this should be moved into the `putOrUpdateTransformStats` listener AND that the state that is saved with that listener is `STOPPED` for both the indexer and the task. 

I am not sure we want to change the state of the task and the indexer to STOPPED inside memory, but I do think it should be persisted as such, and then call shutdown.

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameUsageIT.java [hunk 2]
```diff
@@ -105,8 +105,8 @@ public void testUsage() throws Exception {
         usageAsMap = entityAsMap(usageResponse);
         // we should see some stats
         assertEquals(3, XContentMapValues.extractValue("data_frame.transforms._all", usageAsMap));
-        assertEquals(1, XContentMapValues.extractValue("data_frame.transforms.started", usageAsMap));
-        assertEquals(2, XContentMapValues.extractValue("data_frame.transforms.stopped", usageAsMap));
+        // TODO: due to auto-stop we only see stopped data frames
+        assertEquals(3, XContentMapValues.extractValue("data_frame.transforms.stopped", usageAsMap));
```
Comment (benwtrent) [inline x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameUsageIT.java]: It seems to me that there may be a runtime problem where the single task that was started has not actually ticked over to `STOPPED` by the time this method is called, maybe an `assertBusy`?

Comment (hendrikmuhs) [reply x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameUsageIT.java]: `startAndWaitForTransform` called on line 70 does the `assertBusy`, see changes in `DataFrameRestTestCase`

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java [hunk 3]
```diff
@@ -575,6 +587,10 @@ protected void doSaveState(IndexerState indexerState, Map<String, Object> positi
                             DataFrameTransformCheckpointingInfo.EMPTY), // TODO should this be null
                     ActionListener.wrap(
                             r -> {
+                                // for auto stop shutdown the task
+                                if (state.getTaskState().equals(DataFrameTransformTaskState.STOPPED)) {
+                                    onStop();
+                                }
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: I am unsure whether next.run() should be called 1st or 2nd. Right now it does not matter as it's an empty lambda for this case

**Review by benwtrent — APPROVED** [review]

**Pull request merged**